Sources/_RegexParser/Regex/Parse/LexicalAnalysis.swift

```diff
@@ -16,178 +16,259 @@ concerns upon request.
-  ///
-  /// Note: much of the time, but not always, we can vend a more specific error.
+  /// Diagnoses an error and returns `false` if the end of input has been
+  /// reached. Otherwise returns `true`.
+  @discardableResult
   mutating func expectNonEmpty(
     _ error: ParseError = .unexpectedEndOfInput
-  ) throws {
-    _ = try recordLoc { src in
-      if src.isEmpty { throw error }
+  ) -> Bool {
+    guard !src.isEmpty else {
+      errorAtCurrentPosition(error)
+      return false
     }
+    return true
   }
 
-  mutating func tryEatNonEmpty<C: Collection>(sequence c: C) throws -> Bool
-    where C.Element == Char
-  {
-    try expectNonEmpty(.expected(String(c)))
-    return tryEat(sequence: c)
+  /// Attempt to eat a sequence of characters, additionally diagnosing if the
+  /// end of the source has been reached.
+  mutating func tryEatNonEmpty<C: Collection>(
+    sequence c: C
+  ) -> Bool where C.Element == Char {
+    expectNonEmpty(.expected(String(c))) && tryEat(sequence: c)
   }
 
-  mutating func tryEatNonEmpty(_ c: Char) throws -> Bool {
-    try tryEatNonEmpty(sequence: String(c))
+  /// Returns the next character, or `nil` if the end of the source has been
+  /// reached.
+  func peek() -> Char? { src.peek() }
+
+  /// Same as `peek()`, but with the source location of the next character.
+  func peekWithLoc() -> Located<Char>? {
+    peek().map { c in
+      let nextPos = src.input.index(after: src.currentPosition)
+      return Located(c, Location(src.currentPosition ..< nextPos))
+    }
   }
 
-  /// Attempt to make a series of lexing steps in `body`, returning `nil` if
-  /// unsuccesful, which will revert the source back to its previous state. If
-  /// an error is thrown, the source will not be reverted.
-  mutating func tryEating<T>(
-    _ body: (inout Source) throws -> T?
-  ) rethrows -> T? {
-    // We don't revert the source if an error is thrown, as it's useful to
-    // maintain the source location in that case.
-    let current = self
-    guard let result = try body(&self) else {
-      self = current
-      return nil
+  /// Advance the input `n` characters ahead.
+  mutating func advance(_ n: Int = 1) {
+    guard src.tryAdvance(n) else {
+      unreachable("Advancing beyond end!")
+
+      // Empty out the remaining characters.
+      src.tryAdvance(src._slice.count)
+      return
     }
-    return result
   }
 
-  /// Perform a lookahead using a temporary source. Within the body of the
-  /// lookahead, any modifications to the source will not be reflected outside
-  /// the body.
-  func lookahead<T>(_ body: (inout Source) throws -> T) rethrows -> T {
-    var src = self
-    return try body(&src)
+  /// Eat a character from the input, emitting an unreachable error if the end
+  /// has been reached.
+  mutating func eat() -> Char {
+    guard let char = tryEat() else {
+      unreachable("Cannot eat at the end!")
+      return "\u{FFFD}" // U+FFFD Replacement Character (�)
+    }
```

Comment:
@milseman We used to `assert` and `unsafelyUnwrapped` here. In the spirit of trying not to crash where we can (ala `throw Unreachable`), I changed it to record an `unreachable` error and return U+FFFD. Does that make sense?

Reply:
Would the caller just crash downstream, or what would happen?

Reply:
It would be up to the client, it gets added as a `.fatalError` to the diagnostics, and e.g for the existing clients that expect an error to be thrown (such as the runtime `Regex` initializer), it gets materialized as a thrown error (similar to `Unreachable`). It could be argued that the compiler ought to crash if it gets handed back a `.fatalError`, though we could also emit a `"please file a bug report"` diagnostic.

Reply:
AFAICT, the call to `eat()` is guarded by a source not empty, so it could be changed to just call `tryEat()`. The site that uses `eatWithLoc()` seems like returning `U+FFFD` would fall through to a default switch case, but arguably should be properly handled.

What do you think about just removing `eat()`? Callers can use `tryEat()` and if they want a substitute they can supply it explicitly.

Reply:
> What do you think about just removing `eat()`? Callers can use `tryEat()` and if they want a substitute they can supply it explicitly.

Seems reasonable to me! Updated to remove it